components/butterfly/src/member.rs

```diff
@@ -678,58 +731,67 @@ impl MemberList {
         }
     }
 
-    /// Iterates over every suspected membership entry, calling the given closure.
-    pub fn with_suspects<F>(&self, mut with_closure: F) -> ()
-    where
-        F: FnMut((&str, &SteadyTime)) -> (),
-    {
-        for (id, suspect) in self
-            .suspect
-            .read()
-            .expect("Suspect list lock is poisoned")
-            .iter()
+    // TODO (CM): Better return type than Vec<String>
+    /// Query the list of aging Suspect members to find those which have now
+    /// timed out to Confirmed. Health is updated appropriately, and a
+    /// list of newly-Confirmed Member IDs is returned.
+    pub fn get_newly_expired_confirmed(&self, timeout: Duration) -> Vec<String> {
+        let now = SteadyTime::now();
+        let mut newly_confirmed: Vec<String> = Vec::new();
+
         {
-            with_closure((id, suspect));
+            let mut aging_suspects = self
+                .aging_suspects
+                .write()
+                .expect("aging_suspects lock is poisoned");
+
+            aging_suspects.retain(|ref member_id, ref suspect_time| {
+                let timed_out = now >= **suspect_time + timeout;
+                if timed_out {
+                    newly_confirmed.push(member_id.to_string());
+                }
+                !timed_out
+            });
         }
-    }
 
-    /// Iterates over every suspected membership entry, calling the given closure.
-    pub fn with_departures<F>(&self, mut with_closure: F) -> ()
-    where
-        F: FnMut((&str, &SteadyTime)) -> (),
-    {
-        for (id, departure_time) in self
-            .depart
-            .read()
-            .expect("Departure list lock is poisoned")
-            .iter()
-        {
-            with_closure((id, departure_time));
+        // Since these have "graduated" from Suspect to Confirmed, we
+        // should record their new health.
+        for id in newly_confirmed.iter() {
+            self.insert_health_by_id(id, Health::Confirmed);
         }
-    }
 
-    /// Expires a member from the suspect list.
-    pub fn expire(&self, member_id: &str) {
-        let mut suspects = self.suspect.write().expect("Suspect list lock is poisoned");
-        suspects.remove(member_id);
+        newly_confirmed
     }
 
-    /// Sets a departure time for a member who has been confirmed
-    pub fn depart(&self, member_id: &str) {
-        let mut depart = self
-            .depart
-            .write()
-            .expect("Departure list lock is poisoned");
-        depart.insert(member_id.to_string(), SteadyTime::now());
-    }
+    /// Query the list of aging Confirmed members to find those which have now
+    /// timed out to Departed. Health is updated appropriately, and a
+    /// list of newly-Departed Member IDs is returned.
+    pub fn get_newly_expired_departed(&self, timeout: Duration) -> Vec<String> {
```

Comment:
This seems essentially identical to `get_newly_expired_confirmed`. Why not parameterize the `HashMap` and make this one function?